### PR TITLE
L2geth rollback

### PR DESCRIPTION
**Pull request by curryxbo**

### CODE REVIEW

File: l2geth/rollup/dump/constants.go [hunk 1]
```diff
@@ -9,4 +9,4 @@ var BvmFeeWallet = common.HexToAddress("0x42000000000000000000000000000000000000
 var TssRewardAddress = common.HexToAddress("0x4200000000000000000000000000000000000020")
 var DeadAddress = common.HexToAddress("0xdeaDDeADDEaDdeaDdEAddEADDEAdDeadDEADDEaD")
 var BvmWhitelistAddress = common.HexToAddress("0x4200000000000000000000000000000000000002")
-var BvmReorgAddress = common.HexToAddress("0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD2222")
+var BvmRollbackAddress = common.HexToAddress("0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD2222")
```
Comment (silmaj) [inline l2geth/rollup/dump/constants.go:12]: is this necessary to have this type of address? i think rollback logic will have no transaction to execute, it only triggered by some event context

File: l2geth/core/vm/evm.go [hunk 2]
```diff
@@ -196,9 +195,7 @@ func (evm *EVM) Interpreter() Interpreter {
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
-	//TODO test roll back
-	if addr == dump.BvmReorgAddress {
-		log.Info("Reorg Message:", "input", input)
+	if addr == dump.BvmRollbackAddress {
```
Comment (silmaj) [inline l2geth/core/vm/evm.go:198]: do we need to modify EVM call? and when does this address be referred

File: l2geth/rollup/sync_service.go [hunk 3]
```diff
@@ -1217,8 +1302,24 @@ func (s *SyncService) syncQueueTransactionRange(start, end uint64) error {
 		if err != nil {
 			return fmt.Errorf("Canot get enqueue transaction; %w", err)
 		}
+		log.Info("current block:", "number", s.bc.CurrentHeader().Number.Uint64())
+		hexRawTx := hexutil.Encode(tx.GetMeta().RawTransaction)
+		if len(hexRawTx) > 402 {
```
Comment (silmaj) [inline l2geth/rollup/sync_service.go:1306]: how is this 402 comes out?

**Review by silmaj — DISMISSED** [review]

**Review by silmaj — DISMISSED** [review]

**Review by byteflyfunny — DISMISSED** [review]

File: l2geth/contracts/message/message.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"github.com/mantlenetworkio/mantle/l2geth/accounts/abi"
+	"github.com/mantlenetworkio/mantle/l2geth/common"
+	"math/big"
+	"strings"
+)
+
+var (
+	messageABI *abi.ABI
+	sccABI     *abi.ABI
```
Comment (HaoyangLiu) [inline l2geth/contracts/message/message.go:13]: Why do we need these two global variables?

File: l2geth/contracts/message/message.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package message
+
+import (
+	"fmt"
+	"github.com/mantlenetworkio/mantle/l2geth/accounts/abi"
+	"github.com/mantlenetworkio/mantle/l2geth/common"
+	"math/big"
+	"strings"
```
Comment (HaoyangLiu) [inline l2geth/contracts/message/message.go:8]: Please go format imports

File: l2geth/contracts/message/message_test.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package message
+
+import (
+	"github.com/mantlenetworkio/mantle/l2geth/common/hexutil"
+	"log"
+	"testing"
+)
```
Comment (HaoyangLiu) [inline l2geth/contracts/message/message_test.go:7]: Please go format imports

File: l2geth/eth/gasprice/rollup_gasprice.go [hunk 7]
```diff
@@ -123,3 +126,19 @@ func (gpo *RollupOracle) SetCharge(charge *big.Int) error {
 	log.Info("Set charge", "charge", charge)
 	return nil
 }
+
+// SetCharge sets the charge value held in the BVM_GasPriceOracle
+func (gpo *RollupOracle) SetSCCAddress(sccAddress common.Address) error {
+	gpo.sccAddressLock.Lock()
+	defer gpo.sccAddressLock.Unlock()
+	gpo.sccAddress = sccAddress
+	log.Info("Set sccAddress", "sccAddress", sccAddress)
```
Comment (HaoyangLiu) [inline l2geth/eth/gasprice/rollup_gasprice.go:135]: `log.Info("Set sccAddress", "sccAddress", sccAddress)` -> `log.Info("Set sccAddress", "sccAddress", sccAddress.toString())`

File: l2geth/eth/gasprice/rollup_gasprice.go [hunk 8]
```diff
@@ -123,3 +126,19 @@ func (gpo *RollupOracle) SetCharge(charge *big.Int) error {
...
+	return nil
+}
+
+// SCCAddress returns the cached SCCAddress value
+func (gpo *RollupOracle) SCCAddress() (common.Address, error) {
```
Comment (HaoyangLiu) [inline l2geth/eth/gasprice/rollup_gasprice.go:140]: `error` in return parameter list is not necessary

**Review by HaoyangLiu — APPROVED** [review]

**Review by byteflyfunny — APPROVED** [review]

**Review by wukongcheng — APPROVED** [review]

**Pull request merged**